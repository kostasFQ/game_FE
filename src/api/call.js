exports.makeCall = async (values) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve(values);
    }, 500);
  })
}