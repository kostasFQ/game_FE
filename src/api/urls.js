const url = () => {
  return process.env.REACT_APP_DEV ?
    process.env.REACT_APP_API_URL_DEV :
    process.env.REACT_APP_API_URL_PROD;
}

export const saveResult = () => `${url()}/save`