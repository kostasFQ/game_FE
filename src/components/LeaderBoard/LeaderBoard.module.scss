.leaderBoard {
  &__table {
    width: 80%;
    text-align: center;
    color: rgba(119, 115, 115, 0.7);
    font-size: 0.8em;
    border-collapse: collapse;
    border: solid 1px transparent;

    &__average {
      color: rgba(255, 0, 0, 0.548);
    }

    &__div {
      border-top: solid red 1px;
      border-bottom: solid red 1px;
      background-color: rgba(250, 158, 158, 0.336);
    }
  }
}