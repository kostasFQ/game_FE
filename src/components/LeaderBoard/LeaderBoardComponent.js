import React, { Fragment, PureComponent } from 'react';
import cn from 'classnames';
import styles from './LeaderBoard.module.scss';

class LeaderBoard extends PureComponent {
  state = { list: [], tableHeaders: [], loading: false };

  componentDidMount() {
    this.getLeaders();
  }

  getLeaders = async () => {
    try {
      const { size, getLeaders } = this.props;
      this.setState(() => ({ loading: true }));
      const { data: { response: list } } = await getLeaders(size);
      this.setState(() => ({ loading: false, list }));
    } catch (err) {
      throw new Error(err.message);
    }
  }

  render() {
    const { list, loading } = this.state;
    const { game: { userPlace, userName } } = this.props;
    if (loading) {
      return (
        <h3>Loading...</h3>
      )
    }

    if (list.length !== 0) {
      return (
        <Fragment>
          <table className={styles.leaderBoard__table} border='1'>
            <caption><h3>TOP 10</h3></caption>
            <thead>
              <tr >
                <th>place</th>
                <th>name</th>
                <th>score</th>
                <th className={styles.leaderBoard__table__average}>average</th>
                <th>seconds</th>
              </tr>
            </thead>
            <tbody>
              {
                list.map((i, index) => (
                  <tr key={i.name} className={cn(i.name === userName && styles.leaderBoard__table__div)}>
                    <td>{index + 1}</td>
                    <td>{i.name}</td>
                    <td>{i.score}</td>
                    <td className={styles.leaderBoard__table__average}>{i.average}</td>
                    <td>{i.seconds}</td>
                  </tr>
                ))
              }
            </tbody>
          </table>
          {
            userPlace &&
            <div className={styles.leaderBoard__table__userPlace}>
              {userName}, your place is <span className={styles.leaderBoard__table__average}>{userPlace}</span>
            </div>
          }
        </Fragment>
      )
    }

    return (
      <Fragment>no data</Fragment>
    )

  }
}

export default LeaderBoard;