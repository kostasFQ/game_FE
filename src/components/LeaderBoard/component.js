import React, { Fragment, PureComponent } from 'react';
import { leaderBoard } from 'api/urls';
import cn from 'classnames';
import styles from './LeaderBoard.module.scss';

class LeaderBoard extends PureComponent {
  state = { list: [], tableHeaders: [], loading: false }

  async componentDidMount() {
    this.getLeaders();
  }

  getLeaders = async () => {
    try {
      const { makeCall, size } = this.props;
      this.setState(() => ({ loading: true }));
      const { data: { response: list } } = await makeCall(leaderBoard(size));
      if (list.length !== 0) {
        const tableHeaders = Object.keys(list[0]).slice(1, 5);
        tableHeaders.unshift('place');
        this.setState(() => ({ list, tableHeaders}));
      }
      this.setState(() => ({ loading: false }));
    } catch (err) {
      throw new Error(err.message);
    }
  }

  render() {
    const { list, tableHeaders, loading } = this.state;
    const { game: { userPlace, userName } } = this.props;

    if (loading) {
      return (
        <h3>Loading...</h3>
      )
    }

    return (
      <Fragment>
        <table className={styles.leaderBoard__table} border='1'>
          <caption><h3>{ list.length !== 0 && 'TOP 10' }</h3></caption>
          <thead>
            <tr >
              {tableHeaders.map(i => (<th key={i} className={cn(i === 'average' && styles.leaderBoard__table__average)} >{i.toUpperCase()}</th>))}
            </tr>
          </thead>
          <tbody>
            {
              list.map((i, index) => (
                <tr key={i.name} className={ cn(i.name === userName && styles.leaderBoard__table__div )}>
                  <td>{index + 1}</td>
                  <td>{i.name}</td>
                  <td>{i.score}</td>
                  <td className={styles.leaderBoard__table__average}>{i.average}</td>
                  <td>{i.seconds}</td>
                </tr>
              ))
            }
          </tbody>
        </table>
        {
          userPlace &&
          <div className={styles.leaderBoard__table__userPlace}>
            {userName}, your place is <span className={styles.leaderBoard__table__average}>{userPlace}</span>
          </div>
        }
      </Fragment>
    )
  }
}

export default LeaderBoard;