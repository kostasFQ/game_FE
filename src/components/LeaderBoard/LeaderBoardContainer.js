import { connect } from 'react-redux';
import LeaderBoard from './LeaderBoardComponent';
import { leaderBoard } from 'api/urls';
import makeCall from 'api/call';

const mapDispatchToProps = () => ({
  getLeaders: size => makeCall(leaderBoard(size))
});

const mapStateToProps = state => ({
  game: state.game,
});

export default connect(mapStateToProps, mapDispatchToProps)(LeaderBoard);