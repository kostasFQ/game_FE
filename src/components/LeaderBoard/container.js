import { connect } from 'react-redux';
import LeaderBoard from './component';
import makeCall from 'api/call';

const mapDispatchToProps = () => ({
  makeCall: url => makeCall(url)
});


const mapStateToProps = state => ({
  game: state.game,
});

export default connect(mapStateToProps, mapDispatchToProps)(LeaderBoard);