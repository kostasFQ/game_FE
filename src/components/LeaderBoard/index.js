import LeaderBoard from './LeaderBoardContainer';

export default LeaderBoard;