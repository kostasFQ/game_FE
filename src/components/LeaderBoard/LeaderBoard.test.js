import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { StaticRouter } from 'react-router-dom';
import _store from '../../../mocks/store';
import LeaderBoard from './index';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { leaderBoard } from 'api/urls';

const mockStore = configureStore([thunk]);
const initialState = { game: _store };
const store = mockStore(initialState);

test('LeaderBoard should be render', async () => {
  const instance = await renderer.create(
    <Provider store={store}>
      <StaticRouter location="/game">
        <LeaderBoard />
      </StaticRouter>
    </Provider>,
  );
  let component = instance.toJSON();
  expect(component).toMatchSnapshot();
});

test('LeaderBoard get data', async () => {
  const instance = await renderer.create(
    <Provider store={store}>
      <StaticRouter location="/game">
        <LeaderBoard />
      </StaticRouter>
    </Provider>,
  );

  const time = () => new Promise((res, rej) => {
    setTimeout(() => {
      res("result");
    }, 1000)
  })
  await time();

  let component = instance.toJSON();
  expect(component).toMatchSnapshot();
});

test('leader board error', async () => {
  const mock = new MockAdapter(axios);
  mock.onGet(leaderBoard(10)).networkError();

  const instance = await renderer.create(
    <Provider store={store}>
      <StaticRouter location="/game">
        <LeaderBoard />
      </StaticRouter>
    </Provider>,
  );

  let component = instance.toJSON();
  expect(component).toMatchSnapshot();
  mock.reset();
});

