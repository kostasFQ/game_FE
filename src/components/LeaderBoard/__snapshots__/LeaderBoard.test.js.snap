// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderBoard get data 1`] = `
Array [
  <table
    border="1"
    className="leaderBoard__table"
  >
    <caption>
      <h3>
        TOP 10
      </h3>
    </caption>
    <thead>
      <tr>
        <th>
          place
        </th>
        <th>
          name
        </th>
        <th>
          score
        </th>
        <th
          className="leaderBoard__table__average"
        >
          average
        </th>
        <th>
          seconds
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className=""
      >
        <td>
          1
        </td>
        <td>
          kost
        </td>
        <td>
          50
        </td>
        <td
          className="leaderBoard__table__average"
        >
          10
        </td>
        <td>
          5
        </td>
      </tr>
      <tr
        className=""
      >
        <td>
          2
        </td>
        <td>
          2
        </td>
        <td>
          45
        </td>
        <td
          className="leaderBoard__table__average"
        >
          9
        </td>
        <td>
          5
        </td>
      </tr>
      <tr
        className=""
      >
        <td>
          3
        </td>
        <td>
          kst1
        </td>
        <td>
          40
        </td>
        <td
          className="leaderBoard__table__average"
        >
          8
        </td>
        <td>
          5
        </td>
      </tr>
      <tr
        className=""
      >
        <td>
          4
        </td>
        <td>
          kst2
        </td>
        <td>
          40
        </td>
        <td
          className="leaderBoard__table__average"
        >
          8
        </td>
        <td>
          5
        </td>
      </tr>
      <tr
        className="leaderBoard__table__div"
      >
        <td>
          5
        </td>
        <td>
          kst
        </td>
        <td>
          35
        </td>
        <td
          className="leaderBoard__table__average"
        >
          7
        </td>
        <td>
          5
        </td>
      </tr>
    </tbody>
  </table>,
  <div
    className="leaderBoard__table__userPlace"
  >
    kst
    , your place is 
    <span
      className="leaderBoard__table__average"
    >
      1
    </span>
  </div>,
]
`;

exports[`LeaderBoard should be render 1`] = `
<h3>
  Loading...
</h3>
`;

exports[`leader board error 1`] = `
<h3>
  Loading...
</h3>
`;
